### PR TITLE
074 dirt texture erasing

### PR DESCRIPTION
**Pull request by CrispyKorn**
Feature Components:
- Create tool scriptable objects and use any sprite for it's erasing mask.
- Erase textures using any tool.
- Adjust the size and strength of each tool from within the scriptable object.
- Duplicate any erasable texture as much as you like.

Note: Objects who's sprite you want to be erasable must be tagged 'Erasable'.

📝 Checklist before requesting a review:
- [x]  I have run this branch locally
- [x]  I have performed a self-review of my code
- [x]  I have confirmed critical features still function

### CODE REVIEW

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/Eraser.cs:99]: Texture2D.alphaIsTransparency is editor only and can't be changed unless done in TextureImporter.alphaIsTransparency. Also it breaks exporting builds

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/Eraser.cs:118]: Would be good to have this follow the hand instead of the mouse so we don't have 2 different controls

File: MonstieWash/Assets/Scripts/FollowMouse.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/FollowMouse.cs:20]: same as other comment to follow hand instead of mouse

File: MonstieWash/Assets/Scripts/Tool.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/Tool.cs:22]: 1.0472 should probably be explained,  eg. scaling factor, also does inputStrength need to be adjusted

File: MonstieWash/Assets/Scripts/Tool.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/Tool.cs:14]: Could separate what is/isn't shown in inspector

File: MonstieWash/Assets/Scenes/MonstieWash.unity
Comment (SkyRiser568) [inline MonstieWash/Assets/Scenes/MonstieWash.unity:512]: Dirt object is untagged
should be Erasable?
```suggestion
  m_TagString: Erasable
```

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/Eraser.cs:65]: If we'll need to document our functions with XML tags it would be nice to have it as part of the conventions in the dev wiki

**Review by SkyRiser568 — CHANGES_REQUESTED** [review]
Great stuff! like WOW. 
Line 99 has something that causes error when trying to export a build.
Then there's some other changes/cleanup for you to go over.
😄

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/Eraser.cs:99]: Hmm that's v unfortunate. I'll have a look at another way to do this that is legal.

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/Eraser.cs:118]: Ah yep most of this was done before your feature was added just for testing, I can update it to include your work. 👍

File: MonstieWash/Assets/Scripts/Tool.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/Tool.cs:22]: Ah good catch, will do. I'll explain more of what each variable does in the code (the 1.0472 is just 100 to the 100th root), but for now the inputStrength is what the user inputs on a slider (0 - 100) which represents the strength/opacity of the tool and then that is scaled to make the strength adjustments seem more linear and intuitive. Then strength is the final (actual) tool strength value. These are mainly settings for designers to use when setting up tools.

File: MonstieWash/Assets/Scripts/Tool.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/Tool.cs:14]: Will do, good idea! I thought that was messy but wasn't sure the best way to lay it out.

File: MonstieWash/Assets/Scenes/MonstieWash.unity
Comment (CrispyKorn) [reply MonstieWash/Assets/Scenes/MonstieWash.unity:512]: Oh this was just for demonstration of how the tag works, these objects aren't important (just for testing) and will most likely be deleted anyways.

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/Eraser.cs:65]: Ah yep, good idea. I think I'll add it there as it's v helpful when reading through the code.

**Comment by CrispyKorn** [comment]
I've fixed up the code based on your feedback and made my system work with yours. Also, turns out I didn't even need the line that broke the build after all .-. so I just removed it. Take another look and again let me know if you'd like anything changed!

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/PlayerHand.cs:41]: thanks for this, way cleaner method that I forgot about

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/PlayerHand.cs:35]: I think you could use Camera.aspect here for if the aspect ratio is changed

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/PlayerHand.cs:26]: This one is my bad, handPosition should probably be assigned in world point because it will need to be compared to many other objects in scene which also default to world point. I only changed it to screen point because I had to compare to screen edges

File: MonstieWash/Assets/Scripts/FollowMouse.cs
Comment (SkyRiser568) [inline MonstieWash/Assets/Scripts/FollowMouse.cs:1]: I don't think this script is being used anymore or I just didn't see where. If it's deprecated it would be good to leave a note in the script for why.

**Review by SkyRiser568 — CHANGES_REQUESTED** [review]
Changes are good, there are a few bits of cleanup I think would be good before merging.

File: MonstieWash/Assets/Scripts/Eraser.cs
Comment (SkyRiser568) [reply MonstieWash/Assets/Scripts/Eraser.cs:3]: this should just be about line 3 with using UnityEditor

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/PlayerHand.cs:26]: Ah I see, well in that case hand position can be found just from transform.position, so would we want to still keep handPosition at all? I use it currently in the eraser code but I don't necessarily need it.

File: MonstieWash/Assets/Scripts/FollowMouse.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/FollowMouse.cs:1]: Yep it's from before your code was done, I've just deleted it since it's not used for anything anymore.

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (CrispyKorn) [reply MonstieWash/Assets/Scripts/PlayerHand.cs:35]: Good catch!

File: MonstieWash/Assets/Scripts/PlayerHand.cs
Comment (SkyRiser568) [reply MonstieWash/Assets/Scripts/PlayerHand.cs:26]: i'd remove it since its not needed

**Comment by SkyRiser568** [comment]
Since these changes should be simple, if there's nothing new that needs review after you're done. I reckon you can just merge it to dev

**Comment by CrispyKorn** [comment]
Actually I just added the Input System so if you could take a look at that just to make sure it's all okay (it all works great in testing) that would be great! After that we should be good to merge.

**Comment by CrispyKorn** [comment]
Also if possible it'd be great if you could create a new "approve" review when you're happy with it just for the sake of the process (which I will update on notion when I get the chance) and so the status of the pr is updated. Thanks!!

**Review by SkyRiser568 — APPROVED** [review]
The requested changes are all good. Merge can be done after looking at the comments.

I think changing the input system now was a bit out of scope for this feature and premature, I would have liked some input from other programmers before changing systems and maybe even from the designers because it will impact them down the line. 
I don't think its worth reverting anything that has been done but we should schedule a meeting for it.

**Comment by CrispyKorn** [comment]
> The requested changes are all good. Merge can be done after looking at the comments.
> 
> I think changing the input system now was a bit out of scope for this feature and premature, I would have liked some input from other programmers before changing systems and maybe even from the designers because it will impact them down the line. I don't think its worth reverting anything that has been done but we should schedule a meeting for it.

I understand, I thought it was a given we would use the Input System as that's the proper way to do controls nowadays but you're right that it was out of scope for this feature, I just wanted to get it out of the way early. I'll change how its implemented and merge the changes. Thanks for your patience!

**Pull request merged**
